judoscale/celery/collector.py

```diff
@@ -0,0 +1,60 @@
+import json
+import logging
+from typing import List, Optional, Set
+
+from celery import Celery
+from redis import Redis
+
+from judoscale.core.config import Config
+from judoscale.core.metric import Metric
+from judoscale.core.metrics_collectors import JobMetricsCollector
+
+logger = logging.getLogger(__name__)
+
+
+class CeleryMetricsCollector(JobMetricsCollector):
+    def __init__(self, config: Config, broker: Celery):
+        connection = broker.connection_for_read()
+        if connection.transport.driver_name != "redis":
+            raise NotImplementedError(
+                f"Unsupported broker: {connection.transport.driver_name}"
+            )
+
+        self.redis: Redis = connection.channel().client
+        logger.debug(f"Redis is at {self.redis.connection_pool.connection_kwargs}")
+        super().__init__(config=config)
+
+    @property
+    def queues(self) -> Set[str]:
+        """
+        Get all queues from Redis.
+        """
+        system_queues = {"unacked", "unacked_index"}
+        queues = set(self.redis.scan_iter(match="[^_]*", _type="list"))
```

Comment:
It looks like this is run every time `collect` is run, is that right? I'm definitely concerned about hammering Redis if we have a bunch of Celery processes doing this every few seconds.

> A more efficient way would be to use the [active_queues()](https://docs.celeryq.dev/en/stable/reference/celery.app.control.html#celery.app.control.Inspect.active_queues) method from celery.app.control.Inspect, but this only works if there are workers running, because workers register which queues they will be consuming from.

Ah, and I suppose this would be a problem if worker dynos are scaled down to zero and we're relying on web dynos to report Celery queue time. Is that what you were thinking as well?

Another idea would be to only run this once when the collector is instantiated for an initial "snapshot" of the known queues. This would miss queues that are added after the process has been started, so maybe we also run it periodically, but less frequently than `collect` (maybe every 10th call to `collect` or something like that). Feels messy and I don't love it... just spitballing ideas.

Another idea is if `before_publish` had access to the collector, perhaps it could add the queue to the `collector.queues` set. I would think this would be a very fast operation since 99% of the time it would be a no-op (the queue is already present in the set).

Reply:
> Ah, and I suppose this would be a problem if worker dynos are scaled down to zero and we're relying on web dynos to report Celery queue time. Is that what you were thinking as well?

That's right. With Celery, the expectation seems to be that there's always at least one worker if you want to use the `Inspect` API.

> Another idea would be to only run this once when the collector is instantiated for an initial "snapshot" of the known queues.

The problem with this is that not all queues may be present when the snapshot is taken. Redis will remove the queue (key) if it's empty.

```sh
127.0.0.1:6379> keys "test_queue"
(empty array)
127.0.0.1:6379> lpush "test_queue" "test"
(integer) 1
127.0.0.1:6379> keys "test_queue"
1) "test_queue"
127.0.0.1:6379> lpop "test_queue" 1
1) "test"
127.0.0.1:6379> keys "test_queue"
(empty array)
127.0.0.1:6379>
```

If there isn't any work going on when the snapshot is taken, the set of queues reported back to the collector could be a subset (perhaps an empty set) of the actual queues the workers will see in production.

> Another idea is if before_publish had access to the collector, perhaps it could add the queue to the collector.queues set. I would think this would be a very fast operation since 99% of the time it would be a no-op (the queue is already present in the set).

I like this idea. The only rub with it is that because Celery allows enqueuing tasks with a routing key, rather than the queue name, we may not be able to determine which queue the task ends up on. I'll have a look tomorrow if there is a way know the queue name at the point of publishing.

> maybe we also run it periodically, but less frequently than collect (maybe every 10th call to collect or something like that)

This sounds like the most robust for getting an accurate set of queues, even though I'm with you on the messiness argument — doesn't feel super clean to me.

I'll do a bit of digging and come up with some approaches.